### DISCUSSION TITLE
New versions - History.md

### DISCUSSION THREAD

**Discussion by MSP-Greg** [puma/puma#2939] [General]
We have used https://github.com/MSP-Greg/issue-pr-link to help with the assembly of change logs, based on the labels used for PR's.  It also provides markdown links for it.

Just pushed to the repo, the change only allows a PR to be listed in the first matching label.  Labels used in the log are provided in an array in the json info file.  The updated json should look like the below, `["breaking change", "Breaking Changes"]` has been added:
```json
{
  "owner": "puma",
  "repo": "puma",
  "history": "History.md",
  "labels": [ ["breaking change", "Breaking Changes"], ["feature", "Features"], ["bug", "Bugfixes"], ["perf", "Performance"], ["refactor", "Refactor"] ],
  "token": "<your GiHub token>"
}
```

Note that since the code only parses PR's, if maintainers push commits directly, it should be a change that doesn't need to be listed in a changelog/history file.  An example would be CI related changes.

**Comment by nateberkopec** [comment]
> if maintainers push commits directly,

Should we protect the master branch?

**Reply by MSP-Greg** [reply]
Might be a good idea.  The main check that might be helpful is 'Require linear history'.

Not sure how it works, but if it needs to be protected to disable 'Allow force pushes' and 'Allow deletions' (leave unchecked?), that would be a good idea.  I assume no one would deliberately do that, but accidentally...

A quick look, and the last commit I did without a PR was 4bea6967a775, which I think commits of that type are ok?  WDYT?

**Reply by nateberkopec** [reply]
Done!

**Reply by dentarg** [reply]
A good thing that comes from protecting the master branch and doing PRs for every change is that it will be easier to follow all changes. I haven't gotten any notifications (emails) for pushes to master.
